### ISSUE TITLE
The bootstrapping breaks when used as a git dependency

### ISSUE THREAD

**Issue by Nadrieril** [pest-parser/pest#376]
I'd like to use the latest pest from the git repo, but I can't because of the bootstrapping phase.
My Cargo.toml look like
```
pest = { git = "https://github.com/pest-parser/pest" }
pest_derive = { git = "https://github.com/pest-parser/pest" }
```
which should work, but the build panics with 'Bootstrap failed because no bootstrap executable was found. Please run `cargo build --package pest_bootstrap` or `cargo bootstrap` and then try again.'
The error is clearly from the brittle bootstrapping code, but I'm not familiar enough with cargo to know how to fix it. My guess would be to use pest_bootstrap as a library instead of a binary in build.rs but there's probably a good reason it's a binary...

**Comment by CAD97** [comment]
TL;DR bootstrap is hard.

Unfortunately, `pest_meta` can't depend on bootstrap directly for two main reasons:
- This introduces a cyclic dependency (`pest_meta` -> `pest_bootstrap` -> `pest` -> `pest_meta`), and
- We don't want to run the bootstrap step when building from distribution (i.e. crates-io).

A third constraint is that we don't want to put generated files into the repository if we can avoid it. It's bad practice and just asking for both merge conflicts and getting out of date.

-----

I'm thinking about ways to improve the bootstrap step. Cargo doesn't exactly make this easy, either: I wish we could `cargo bootstrap` straight from the buildscript but that causes a recursive lock attempt on the build directory, which causes deadlock.

The current workarounds:
- If you build from crates-io first and then from git, it worked once for me at least, though I have _no idea why_. (If I understand the process right, it shouldn't've.)
- Download the git repository, do `cargo bootstrap` (which is really just an alias for `cargo run --package pest_bootstrap`) and use a directory dependency.
- Do `cargo bootstrap` in a pest directory somewhere and then copy the executable to where the build script is looking for it. If I recall the process correctly, it'd be your `target/debug/pest_bootstrap/`.

If there exists a different, better solution to this problem, then I'd love to hear it. I've just not found one yet. (I'm glad my error message was clear enough, however!)

**Comment by Nadrieril** [comment]
Fair enough, I've been using a local clone instead and it worked well. Thanks for the explanation !

**Comment by tomtau** [comment]
with https://github.com/pest-parser/pest/commit/867da2e381b7f28db9496a5879dac67f559936f1

it should be possible to use pet as a git dependency:
```
pest = { git = "https://github.com/pest-parser/pest.git" }
pest_derive = { git = "https://github.com/pest-parser/pest.git", features = ["not-bootstrap-in-src"]}
```

but note that the compilation time will be higher due to the `cargo` library dependency
